File: java/src/test/java/problems101_200/LinkedListCycle2Test.java
package problems101_200;

import com.google.common.collect.Iterables;
import concepts.SinglyListNode;
import org.junit.Test;
import sample.SampledTest;

import static com.google.common.truth.Truth.assertWithMessage;

public class LinkedListCycle2Test extends SampledTest {
    @Override
    protected String getSampleFilename() {
        return "linked-list-cycle-ii.json";
    }

    @Test
    public void test() {
        for (LinkedListCycle2 solution : LinkedListCycle2.values()) {
            for (LinkedListCycle2Sample sample : getSamples(LinkedListCycle2Sample[].class)) {
                SinglyListNode head = SinglyListNode.of(sample.input.head);
                SinglyListNode result = null;
                if (sample.input.pos > -1) {
                    result = Iterables.get(head, sample.input.pos);
                    Iterables.getLast(head).next = result;
                }
                assertWithMessage(sample.getMessage())
                    .that(solution.detectCycle(head))
                    .isEqualTo(result);
            }
        }
    }
}

File: kotlin/src/test/kotlin/problems101_200/LinkedListCycle2Test.kt
package problems101_200

import com.google.common.collect.Iterables
import com.google.common.truth.Truth.assertWithMessage
import concepts.SinglyListNode
import sample.SampledTest
import sample.component1
import sample.component2
import sample.component3
import sample.getSamples
import kotlin.test.Test

class LinkedListCycle2Test : SampledTest() {
    override fun getSampleFilename(): String = "linked-list-cycle-ii.json"

    @Test
    fun test() =
        LinkedListCycle2.entries.forEach { solution ->
            getSamples<Array<LinkedListCycle2Sample>>().forEach { (input, _, message) ->
                var result: SinglyListNode? = null
                assertWithMessage(message)
                    .that(
                        solution.detectCycle(
                            SinglyListNode.of(*input.head)!!.also {
                                if (input.pos <= -1) {
                                    return@also
                                }
                                result = Iterables.get(it, input.pos)
                                Iterables.getLast(it).next = result
                            },
                        ),
                    ).isEqualTo(result)
            }
        }
}
